### PR TITLE
fix: correctly manage roles according to their scope

### PR DESCRIPTION
**Pull request by shadinaif**
- [x] **chore: add new test requirements**: `deepdiff`: for easy dictionary compare. and `edx-organizations`: needed by `CourseCreator` model

- [x] **feat: cache roles user by user**: Better performance and easier management. And more usability in other functions

- [x] **fix: do not use `org_course_creator_group` in test with `courses_id`**: Tests were assuming that `org_course_creator_group` can be used with courses, which is wrong. Replacing that with `instructor` role

- [x] **refactor: move get_accessible_tenant_ids to roles module**: Better place and avoiding circular import

- [x] **fix: validate course-creator roles**: `org_course_creator_group` and `course_creator_group` can exist in `CourseAccessRole` but still not effective because they don't have the granted `CourseCreator` record

- [x] **fix: unify filtering of staff users by using `check_staff_exist_queryset`**. DRY code. And the ability to have the filter as an optional argument

- [x] **feat: add `include_staff` to `query_param`**: add it to related APIs to include users in lists even when they are staff users

- [x] **feat: special logic to add, update, and delete `org_course_group_creator`**: the logic includes managing records of `CourseCreator` model without contradicting an existing global `course_group_creator` role

- [x] **feat: roles edit restrictions**: set restrictions for roles editting
  - **system staff**: callers with `is_staff=1` or `is_superuser=1` have unrestricted access
  - **tenant-wide-staff**: callers with `staff` tenant-wide have unrestricted access except that they cannot edit `staff` tenant-wide role for other users (including themselves)
  - **tenant-wide-course_creator**: callers with `org_course_creator_group` can only edit course-specific roles for users (including themselves)
  - _**Note: the back-end does not support editing global roles like `course_access_group` and `support`, not even for superusers**_

### CODE REVIEW

**Comment by OmarIthawi** [comment]
> feat: roles delete restrictions: users should be able to delete existing tenant-wide roles for other users unless?????

Unless they're org-wide admins. Org-wide staff/admin (same role different names) acts as owners with equal rights.

**Comment by shadinaif** [comment]
Thank you @OmarIthawi , I've edited the description. Please take a look when you have time

**Comment by OmarIthawi** [comment]
>  feat: roles edit restrictions: set restrictions for roles editting

I think this one has been very well thought and better than anything else we can think of at the moment. Great job. Always very neat!

Let me know when this is ready for another review @shadinaif!!

File: tests/test_helpers/test_roles.py [hunk 1]
```diff
@@ -183,6 +189,8 @@ def test_get_user_course_access_roles(base_data):  # pylint: disable=unused-argu
     expected_result['roles']['staff']['course_limited_access'].append('course-v1:ORG2+2+2')
 
     result = get_user_course_access_roles(user_id)
+    print(result)
+    print(expected_result)
```
Comment (OmarIthawi) [inline tests/test_helpers/test_roles.py]: clean please

File: futurex_openedx_extensions/helpers/roles.py [hunk 2]
```diff
@@ -207,6 +207,40 @@ def get_user_course_access_roles(user_id: int) -> dict:
     }
 
 
+def get_accessible_tenant_ids(user: get_user_model, roles_filter: List[str] | None = None) -> List[int]:
```
Comment (OmarIthawi) [inline futurex_openedx_extensions/helpers/roles.py:257]: Can we use `get_user_course_access_roles`? later on?

**Review by OmarIthawi — APPROVED** [review]
Thanks @shadinaif! I quickly reviewed it and found no issues.

Please share on staging so Ahmed can check it out.

**Pull request merged**